assembl/graphql/idea.py

```diff
@@ -618,6 +619,327 @@ class VideoInput(graphene.InputObjectType):
+        announcement = args.get('announcement')
+        if announcement is not None:
+            announcement_title_entries = announcement.get('title_entries')
+            if len(announcement_title_entries) == 0:
+                raise Exception('Announcement titleEntries needs at least one entry')
+
+            if not thematic.announcement:
+                announcement_title_langstring = langstring_from_input_entries(announcement_title_entries)
+                announcement_body_langstring = langstring_from_input_entries(announcement.get('body_entries', None))
+                saobj2 = create_idea_announcement(user_id, discussion, thematic, announcement_title_langstring, announcement_body_langstring)
+                db.add(saobj2)
+            else:
+                update_langstring_from_input_entries(
+                    thematic.announcement, 'title', announcement_title_entries)
+                update_langstring_from_input_entries(
+                    thematic.announcement, 'body', announcement.get('body_entries', None))
+                thematic.announcement.last_updated_by_id = user_id
+
+        questions_input = args.get('questions')
+        existing_questions = {
+            question.id: question for question in thematic.get_children()}
+        updated_questions = set()
+        if questions_input is not None:
+            for idx, question_input in enumerate(questions_input):
+                if question_input.get('id', None) is not None:
+                    id_ = int(Node.from_global_id(question_input['id'])[1])
+                    updated_questions.add(id_)
+                    question = models.Question.get(id_)
+                    # archive the question
+                    question.copy(tombstone=True)
+                    update_langstring_from_input_entries(
+                        question, 'title', question_input['title_entries'])
+                    # modify question order
+                    question.source_links[0].order = idx + 1.0
+                else:
+                    title_ls = langstring_from_input_entries(
+                        question_input['title_entries'])
+                    question = models.Question(
+                        title=title_ls,
+                        discussion_id=discussion_id
+                    )
+                    db.add(
+                        models.IdeaLink(source=thematic, target=question,
+                                        order=idx + 1.0))
+
+            # remove question (tombstone) that are not in questions_input
+            for question_id in set(existing_questions.keys()
+                                   ).difference(updated_questions):
+                existing_questions[question_id].is_tombstone = True
+        elif not isinstance(thematic, models.Thematic):
+            update_ideas_recursively(thematic, args.get('children', []), phase, context)
+
+    db.flush()
+    return thematic
+
```

Comment:
maybe some parts of `create_idea` and `update_idea` could be factorized (every `if`?) but it's hard to see

Reply:
I don't think so. We use langstring_from_input_entries vs update_langstring_from_input_entries